File: app/src/main/java/com/limelight/binding/audio/MicrophoneManager.java
package com.limelight.binding.audio;

import android.content.Context;
import android.content.pm.PackageManager;
import android.widget.ImageButton;
import android.widget.Toast;

import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import com.limelight.LimeLog;
import com.limelight.R;
import com.limelight.nvstream.NvConnection;
import com.limelight.preferences.PreferenceConfiguration;

/**
 * 麦克风管理器
 * 负责管理麦克风的权限、状态和UI更新
 */
public class MicrophoneManager {
    private static final String TAG = "MicrophoneManager";
    
    private final Context context;
    private final NvConnection connection;
    private MicrophoneStream microphoneStream;
    private ImageButton micButton;
    private boolean enableMic;
    
    public interface MicrophoneStateListener {
        void onMicrophoneStateChanged(boolean isActive);
        void onPermissionRequested();
    }
    
    private MicrophoneStateListener stateListener;
    
    public MicrophoneManager(Context context, NvConnection connection, boolean enableMic) {
        this.context = context;
        this.connection = connection;
        this.enableMic = enableMic;
    }
    
    /**
     * 设置状态监听器
     */
    public void setStateListener(MicrophoneStateListener listener) {
        this.stateListener = listener;
    }
    
    /**
     * 设置麦克风按钮
     */
    public void setMicrophoneButton(ImageButton button) {
        this.micButton = button;
        setupMicrophoneButton();
    }
    
    /**
     * 初始化麦克风流
     */
    public boolean initializeMicrophoneStream() {
        if (!enableMic) {
            LimeLog.info("麦克风功能已禁用");
            return false;
        }
        
        if (microphoneStream != null) {
            LimeLog.info("麦克风流已存在");
            return true;
        }

        if (!hasMicrophonePermission()) {
            showMessage(context.getString(R.string.mic_permission_required));
            return false;
        }
        
        try {
            MicrophoneConfig.updateBitrateFromConfig(context);
            microphoneStream = new MicrophoneStream(connection);

            if (!microphoneStream.start()) {
                showMessage("无法启动麦克风流");
                return false;
            }

            LimeLog.info("麦克风流启动成功");
            
            if (!microphoneStream.isMicrophoneAvailable()) {
                showMessage("主机不支持麦克风功能");
            }

            // 初始化后默认暂停麦克风
            if (microphoneStream.isRunning()) {
                microphoneStream.pause();
                LimeLog.info("麦克风流已初始化，默认状态为关闭");
            }
            
            return true;
        } catch (Exception e) {
            LimeLog.warning("初始化麦克风流失败: " + e.getMessage());
            showMessage("初始化麦克风流失败: " + e.getMessage());
            return false;
        }
    }
    
    /**
     * 切换麦克风状态
     */
    public void toggleMicrophone() {
        if (!checkMicrophonePermission()) return;
        
        if (microphoneStream != null) {
            if (microphoneStream.isRunning()) {
                pauseMicrophone();
            } else {
                resumeMicrophone();
            }
        } else if (connection != null) {
            if (initializeMicrophoneStream()) {
                updateMicrophoneButtonState();
                showMessage(context.getString(R.string.mic_disabled));
            } else {
                showMessage("麦克风状态切换: 初始化失败");
            }
        } else {
            showMessage("麦克风状态切换: 连接不存在");
        }
        
        updateMicrophoneButtonState();
    }
    
    /**
     * 暂停麦克风
     */
    public void pauseMicrophone() {
        if (microphoneStream != null && microphoneStream.isRunning()) {
            microphoneStream.pause();
            showMessage(context.getString(R.string.mic_disabled));
            notifyStateChange(false);
        }
    }
    
    /**
     * 恢复麦克风
     */
    public void resumeMicrophone() {
        if (!checkMicrophonePermission()) return;
        
        if (microphoneStream != null && !microphoneStream.isRunning()) {
            if (microphoneStream.resume()) {
                showMessage(context.getString(R.string.mic_enabled));
                notifyStateChange(true);
            } else {
                restartMicrophoneStream();
            }
        }
    }

    private void restartMicrophoneStream() {
        LimeLog.warning("麦克风恢复失败，尝试重新初始化");
        microphoneStream.stop();
        MicrophoneConfig.updateBitrateFromConfig(context);

        microphoneStream = new MicrophoneStream(connection);
        if (microphoneStream.start()) {
            showMessage(context.getString(R.string.mic_enabled));
            notifyStateChange(true);
        } else {
            showMessage("麦克风恢复失败: 重新初始化失败");
        }
    }

    private void notifyStateChange(boolean isActive) {
        if (stateListener != null) {
            stateListener.onMicrophoneStateChanged(isActive);
        }
    }
    
    /**
     * 获取麦克风当前状态
     */
    public boolean isMicrophoneActive() {
        return microphoneStream != null && microphoneStream.isRunning();
    }
    
    /**
     * 检查麦克风是否可用
     */
    public boolean isMicrophoneAvailable() {
        return microphoneStream != null && microphoneStream.isMicrophoneAvailable();
    }
    
    /**
     * 检查麦克风是否真正可用
     */
    public boolean isMicrophoneTrulyAvailable() {
        return hasMicrophonePermission() &&
                microphoneStream != null &&
                microphoneStream.isMicrophoneAvailable();
    }
    
    /**
     * 检查麦克风权限
     */
    public boolean checkMicrophonePermission() {
        if (!hasMicrophonePermission()) {
            requestMicrophonePermission();
            return false;
        }
        return true;
    }
    
    /**
     * 检查是否有麦克风权限
     */
    public boolean hasMicrophonePermission() {
        return ContextCompat.checkSelfPermission(context,
                android.Manifest.permission.RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED;
    }
    
    /**
     * 请求麦克风权限
     */
    public void requestMicrophonePermission() {
        if (context instanceof android.app.Activity) {
            ActivityCompat.requestPermissions((android.app.Activity) context, 
                    new String[]{android.Manifest.permission.RECORD_AUDIO}, 
                    MicrophoneConfig.PERMISSION_REQUEST_MICROPHONE);
            if (stateListener != null) {
                stateListener.onPermissionRequested();
            }
        }
    }
    
    /**
     * 处理权限请求结果
     */
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        if (requestCode == MicrophoneConfig.PERMISSION_REQUEST_MICROPHONE &&
                grantResults.length > 0) {

            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                new android.os.Handler().postDelayed(() -> {
                    if (hasMicrophonePermission()) {
                        toggleMicrophone();
                    } else {
                        showPermissionError();
                    }
                }, MicrophoneConfig.PERMISSION_DELAY_MS);
            } else {
                showPermissionError();
            }
        }
    }
    
    /**
     * 设置麦克风按钮
     */
    private void setupMicrophoneButton() {
        if (micButton == null) return;

        micButton.setVisibility(enableMic ? android.view.View.VISIBLE : android.view.View.GONE);
        if (enableMic) {
            updateMicrophoneButtonState();
            micButton.setOnClickListener(v -> {
                if (checkMicrophonePermission()) {
                    toggleMicrophone();
                }
            });
        }
    }
    
    /**
     * 更新麦克风按钮状态
     */
    public void updateMicrophoneButtonState() {
        if (micButton == null) return;

        boolean isActive = isMicrophoneActive();
        micButton.setSelected(isActive);

        // 根据配置获取图标资源
        int iconResource = getMicIconResource(isActive);
        micButton.setImageResource(iconResource);

        micButton.setContentDescription(context.getString(
                isActive ? R.string.mic_enabled : R.string.mic_disabled));
        micButton.setEnabled(true);
    }

    /**
     * 根据配置获取麦克风图标资源ID
     * 使用资源映射表，代码更简洁
     */
    private int getMicIconResource(boolean isActive) {
        // 读取配置
        PreferenceConfiguration prefConfig = PreferenceConfiguration.readPreferences(context);
        String colorScheme = prefConfig.micIconColor != null ? prefConfig.micIconColor : "solid_white";
        
        // 资源映射表：简化代码
        String resourceName = isActive ? 
                getEnabledIconResourceName(colorScheme) : 
                getDisabledIconResourceName(colorScheme);
        
        return context.getResources().getIdentifier(
                resourceName, "drawable", context.getPackageName());
    }
    
    /**
     * 获取启用状态的图标资源名称
     */
    private String getEnabledIconResourceName(String colorScheme) {
        switch (colorScheme) {
            case "gradient_blue": return "ic_btn_mic_gradient_blue";
            case "gradient_purple": return "ic_btn_mic_gradient_purple";
            case "gradient_green": return "ic_btn_mic_gradient_green";
            case "gradient_orange": return "ic_btn_mic_gradient_orange";
            case "gradient_red": return "ic_btn_mic_gradient_red";
            default: return "ic_btn_mic";
        }
    }
    
    /**
     * 获取禁用状态的图标资源名称
     */
    private String getDisabledIconResourceName(String colorScheme) {
        switch (colorScheme) {
            case "gradient_blue": return "ic_btn_mic_gradient_blue_disabled";
            case "gradient_purple": return "ic_btn_mic_gradient_purple_disabled";
            case "gradient_green": return "ic_btn_mic_gradient_green_disabled";
            case "gradient_orange": return "ic_btn_mic_gradient_orange_disabled";
            case "gradient_red": return "ic_btn_mic_gradient_red_disabled";
            default: return "ic_btn_mic_disabled";
        }
    }

    /**
     * 停止麦克风流
     */
    public void stopMicrophoneStream() {
        if (microphoneStream != null) {
            microphoneStream.stop();
            microphoneStream = null;
        }
    }
    
    private void showPermissionError() {
        showMessage(context.getString(R.string.mic_permission_required));
    }
    
    private void showMessage(String message) {
        if (context instanceof android.app.Activity) {
            ((android.app.Activity) context).runOnUiThread(() -> {
                Toast.makeText(context, message, Toast.LENGTH_SHORT).show();
            });
        } else {
            // 如果不是 Activity，使用 Handler 确保在主线程中执行
            new android.os.Handler(android.os.Looper.getMainLooper()).post(() -> {
                Toast.makeText(context, message, Toast.LENGTH_SHORT).show();
            });
        }
    }
    
    /**
     * 测试麦克风状态
     */
    public void testMicrophoneStatus() {
        String status = String.format("权限: %s, 流存在: %s, 可用: %s, 激活: %s",
                hasMicrophonePermission(),
                microphoneStream != null,
                isMicrophoneAvailable(),
                isMicrophoneActive());
        
        LimeLog.info("麦克风状态: " + status);
        Toast.makeText(context, status, Toast.LENGTH_LONG).show();
    }
    
    /**
     * 设置麦克风启用状态
     */
    public void setEnableMic(boolean enable) {
        this.enableMic = enable;
        
        if (enable) {
            MicrophoneConfig.updateBitrateFromConfig(context);
        }
        
        if (micButton != null) {
            setupMicrophoneButton();
        }
    }
    
    /**
     * 获取麦克风流实例
     */
    public MicrophoneStream getMicrophoneStream() {
        return microphoneStream;
    }
    
    /**
     * 设置麦克风默认状态为关闭
     */
    public void setDefaultStateOff() {
        if (microphoneStream != null && microphoneStream.isRunning()) {
            microphoneStream.pause();
            LimeLog.info("麦克风已设置为默认关闭状态");
        }
        updateMicrophoneButtonState();
    }
} 

File: app/src/main/java/com/limelight/binding/audio/MicrophoneManager.kt
package com.limelight.binding.audio

import android.content.Context
import android.content.pm.PackageManager
import android.os.Handler
import android.os.Looper
import android.view.View
import android.widget.ImageButton
import android.widget.Toast

import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.limelight.LimeLog
import com.limelight.R
import com.limelight.nvstream.NvConnection
import com.limelight.nvstream.jni.MoonBridge
import com.limelight.preferences.MicrophoneInitialState
import com.limelight.preferences.PreferenceConfiguration

class MicrophoneManager(
    private val context: Context,
    private val connection: NvConnection?,
    private var enableMic: Boolean,
    private val hostUuid: String? = null,
) {
    private var microphoneStream: MicrophoneStream? = null
    private var micButton: ImageButton? = null
    private val buttonPreferences = MicrophoneButtonPreferences(context)
    private val initialStateStore = MicrophoneInitialStateStore(context)
    private var pendingInitialStart = false
    private var lifecycleGeneration = 0L
    private var permissionRequestGeneration: Long? = null
    private var permissionRequestStartsInitial = false

    interface MicrophoneStateListener {
        fun onMicrophoneStateChanged(isActive: Boolean)
        fun onPermissionRequested()
    }

    private var stateListener: MicrophoneStateListener? = null

    fun setStateListener(listener: MicrophoneStateListener?) {
        this.stateListener = listener
    }

    fun setMicrophoneButton(button: ImageButton?) {
        this.micButton = button
        setupMicrophoneButton()
    }

    fun initializeMicrophoneStream(): Boolean {
        if (!enableMic) {
            LimeLog.info("麦克风功能已禁用")
            return false
        }

        val activeConnection = connection ?: run {
            showMessage(context.getString(R.string.mic_error_no_connection))
            return false
        }

        if (microphoneStream != null) {
            LimeLog.info("麦克风流已存在")
            return true
        }

        if (!hasMicrophonePermission()) {
            showMessage(context.getString(R.string.mic_permission_required))
            return false
        }

        try {
            MicrophoneConfig.updateBitrateFromConfig(context)
            MicrophoneConfig.updateVolumeProcessingFromConfig(context)
            microphoneStream = MicrophoneStream(activeConnection)

            if (!microphoneStream!!.start()) {
                microphoneStream?.stop()
                microphoneStream = null
                showMessage(context.getString(R.string.mic_error_start))
                return false
            }

            LimeLog.info("麦克风流启动成功")

            if (!microphoneStream!!.isMicrophoneAvailable()) {
                showMessage(context.getString(R.string.mic_error_host_unsupported))
            }

            if (microphoneStream!!.isRunning()) {
                microphoneStream!!.pause()
                LimeLog.info("麦克风流已初始化，默认状态为关闭")
            }

            return true
        } catch (e: Exception) {
            microphoneStream?.stop()
            microphoneStream = null
            LimeLog.warning(context.getString(R.string.mic_error_init_detail, e.message.orEmpty()))
            showMessage(context.getString(R.string.mic_error_init_detail, e.message.orEmpty()))
            return false
        }
    }

    /** Applies the configured initial state after the stream handshake has completed. */
    fun applyInitialState(initialState: MicrophoneInitialState) {
        if (!enableMic || !MoonBridge.isMicrophoneRequested()) {
            pendingInitialStart = false
            updateMicrophoneButtonState()
            return
        }

        if (!initialState.resolve(initialStateStore.load(hostUuid))) {
            pendingInitialStart = false
            setDefaultStateOff()
            return
        }

        pendingInitialStart = true
        if (!hasMicrophonePermission()) {
            requestMicrophonePermission(forInitialStart = true)
            return
        }
        startPendingInitialState()
    }

    private fun startPendingInitialState() {
        if (!pendingInitialStart) return
        if (!enableMic || !MoonBridge.isMicrophoneRequested() || !hasMicrophonePermission()) {
            pendingInitialStart = false
            updateMicrophoneButtonState()
            return
        }

        if (!initializeMicrophoneStream()) {
            pendingInitialStart = false
            updateMicrophoneButtonState()
            return
        }

        pendingInitialStart = false
        resumeMicrophone(persistState = false)
    }

    fun toggleMicrophone() {
        if (!checkMicrophonePermission()) return

        if (microphoneStream != null) {
            if (microphoneStream!!.isRunning()) {
                pauseMicrophone()
            } else {
                resumeMicrophone()
            }
        } else if (connection != null) {
            if (initializeMicrophoneStream()) {
                resumeMicrophone()
            } else {
                showMessage(context.getString(R.string.mic_error_init))
            }
        } else {
            showMessage(context.getString(R.string.mic_error_no_connection))
        }

        updateMicrophoneButtonState()
    }

    fun pauseMicrophone() {
        if (microphoneStream != null && microphoneStream!!.isRunning()) {
            microphoneStream!!.pause()
            showMessage(context.getString(R.string.mic_disabled))
            notifyStateChange(false)
            updateMicrophoneButtonState()
        }
    }

    fun resumeMicrophone(persistState: Boolean = true) {
        if (!checkMicrophonePermission()) return

        if (microphoneStream != null && !microphoneStream!!.isRunning()) {
            if (microphoneStream!!.resume()) {
                showMessage(context.getString(R.string.mic_enabled))
                notifyStateChange(true, persistState)
                updateMicrophoneButtonState()
            } else {
                restartMicrophoneStream(persistState)
            }
        }
    }

    private fun restartMicrophoneStream(persistState: Boolean) {
        LimeLog.warning("麦克风恢复失败，尝试重新初始化")
        microphoneStream!!.stop()
        MicrophoneConfig.updateBitrateFromConfig(context)
        MicrophoneConfig.updateVolumeProcessingFromConfig(context)

        val activeConnection = connection ?: return
        microphoneStream = MicrophoneStream(activeConnection)
        if (microphoneStream!!.start()) {
            showMessage(context.getString(R.string.mic_enabled))
            notifyStateChange(true, persistState)
            updateMicrophoneButtonState()
        } else {
            showMessage(context.getString(R.string.mic_error_resume))
        }
    }

    private fun notifyStateChange(isActive: Boolean, persistState: Boolean = true) {
        if (persistState) {
            initialStateStore.save(hostUuid, isActive)
        }
        stateListener?.onMicrophoneStateChanged(isActive)
    }

    fun isMicrophoneActive(): Boolean = microphoneStream != null && microphoneStream!!.isRunning()

    fun isMicrophoneAvailable(): Boolean = microphoneStream != null && microphoneStream!!.isMicrophoneAvailable()

    fun isMicrophoneTrulyAvailable(): Boolean {
        return hasMicrophonePermission() &&
                microphoneStream != null &&
                microphoneStream!!.isMicrophoneAvailable()
    }

    fun checkMicrophonePermission(): Boolean {
        if (!hasMicrophonePermission()) {
            requestMicrophonePermission()
            return false
        }
        return true
    }

    fun hasMicrophonePermission(): Boolean {
        return ContextCompat.checkSelfPermission(context,
            android.Manifest.permission.RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED
    }

    fun requestMicrophonePermission(forInitialStart: Boolean = false) {
        permissionRequestGeneration = lifecycleGeneration
        permissionRequestStartsInitial = forInitialStart
        if (context is android.app.Activity) {
            ActivityCompat.requestPermissions(context,
                arrayOf(android.Manifest.permission.RECORD_AUDIO),
                MicrophoneConfig.PERMISSION_REQUEST_MICROPHONE)
            stateListener?.onPermissionRequested()
        }
    }

    fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == MicrophoneConfig.PERMISSION_REQUEST_MICROPHONE &&
            grantResults.isNotEmpty()) {
            val requestGeneration = permissionRequestGeneration ?: return
            val startsInitial = permissionRequestStartsInitial
            permissionRequestGeneration = null
            permissionRequestStartsInitial = false
            if (requestGeneration != lifecycleGeneration) return

            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                Handler(Looper.getMainLooper()).postDelayed({
                    if (requestGeneration != lifecycleGeneration) return@postDelayed
                    if (hasMicrophonePermission()) {
                        if (startsInitial) {
                            if (pendingInitialStart) {
                                startPendingInitialState()
                            }
                        } else {
                            toggleMicrophone()
                        }
                    } else {
                        showPermissionError()
                    }
                }, MicrophoneConfig.PERMISSION_DELAY_MS.toLong())
            } else {
                pendingInitialStart = false
                showPermissionError()
            }
        }
    }

    private fun setupMicrophoneButton() {
        val button = micButton ?: return

        val showButton = enableMic && buttonPreferences.isButtonShownByDefault()

        button.visibility = if (showButton) View.VISIBLE else View.GONE
        if (showButton) {
            updateMicrophoneButtonState()
            button.setOnClickListener {
                if (checkMicrophonePermission()) {
                    toggleMicrophone()
                }
            }
        }
    }

    fun updateMicrophoneButtonState() {
        val button = micButton ?: return

        val isActive = isMicrophoneActive()
        button.isSelected = isActive

        val iconResource = getMicIconResource(isActive)
        button.setImageResource(iconResource)

        button.contentDescription = context.getString(
            if (isActive) R.string.mic_enabled else R.string.mic_disabled
        )
        button.isEnabled = true
    }

    private fun getMicIconResource(isActive: Boolean): Int {
        val prefConfig = PreferenceConfiguration.readPreferences(context)
        val colorScheme = prefConfig.micIconColor ?: "solid_white"

        val resourceName = if (isActive)
            getEnabledIconResourceName(colorScheme)
        else
            getDisabledIconResourceName(colorScheme)

        return context.resources.getIdentifier(resourceName, "drawable", context.packageName)
    }

    private fun getEnabledIconResourceName(colorScheme: String): String {
        return when (colorScheme) {
            "gradient_blue" -> "ic_btn_mic_gradient_blue"
            "gradient_purple" -> "ic_btn_mic_gradient_purple"
            "gradient_green" -> "ic_btn_mic_gradient_green"
            "gradient_orange" -> "ic_btn_mic_gradient_orange"
            "gradient_red" -> "ic_btn_mic_gradient_red"
            else -> "ic_btn_mic"
        }
    }

    private fun getDisabledIconResourceName(colorScheme: String): String {
        return when (colorScheme) {
            "gradient_blue" -> "ic_btn_mic_gradient_blue_disabled"
            "gradient_purple" -> "ic_btn_mic_gradient_purple_disabled"
            "gradient_green" -> "ic_btn_mic_gradient_green_disabled"
            "gradient_orange" -> "ic_btn_mic_gradient_orange_disabled"
            "gradient_red" -> "ic_btn_mic_gradient_red_disabled"
            else -> "ic_btn_mic_disabled"
        }
    }

    fun stopMicrophoneStream() {
        lifecycleGeneration++
        pendingInitialStart = false
        microphoneStream?.stop()
        microphoneStream = null
    }

    private fun showPermissionError() {
        showMessage(context.getString(R.string.mic_permission_required))
    }

    private fun showMessage(message: String) {
        if (context is android.app.Activity) {
            context.runOnUiThread {
                Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
            }
        } else {
            Handler(Looper.getMainLooper()).post {
                Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
            }
        }
    }

    fun testMicrophoneStatus() {
        val status = String.format("权限: %s, 流存在: %s, 可用: %s, 激活: %s",
            hasMicrophonePermission(),
            microphoneStream != null,
            isMicrophoneAvailable(),
            isMicrophoneActive())

        LimeLog.info("麦克风状态: $status")
        Toast.makeText(context, status, Toast.LENGTH_LONG).show()
    }

    fun setEnableMic(enable: Boolean) {
        this.enableMic = enable

        if (enable) {
            MicrophoneConfig.updateBitrateFromConfig(context)
            MicrophoneConfig.updateVolumeProcessingFromConfig(context)
        }

        if (micButton != null) {
            setupMicrophoneButton()
        }
    }

    fun getMicrophoneStream(): MicrophoneStream? = microphoneStream

    fun setDefaultStateOff() {
        if (microphoneStream != null && microphoneStream!!.isRunning()) {
            microphoneStream!!.pause()
            LimeLog.info("麦克风已设置为默认关闭状态")
        }
        updateMicrophoneButtonState()
    }

    companion object {
        private const val TAG = "MicrophoneManager"
    }
}
